### DISCUSSION TITLE
Help with crystal plasticity modeling

### DISCUSSION THREAD

**Discussion by xchengood** [idaholab/moose#19959] [Q&A Modules: Solid mechanics] [answered]
Hi Moose experts or users,

Happy new year.

Recently, I am working on the crystal plasticity model and thinking about how to write specific code for this model (Wen, Wei, Laurent Capolungo, Anirban Patra, and C. N. Tomé. "A physics-based crystallographic modeling framework for describing the thermal creep behavior of Fe-Cr alloys." Metallurgical and Materials Transactions A 48, no. 5 (2017): 2603-2617.) The PDF is available at https://link.springer.com/content/pdf/10.1007/s11661-017-4011-3.pdf.

Two questions:
1. Should I start from equation [1] or equation [4] when writing code?
2. If starting from equation [1], I have no idea about how to express the equations [1] and [2] in Moose. In equation [1], how to express ε_ij and n_i, n_j, b_i and b_j for m_ij. Are there any codes developed for reference?

![image](https://user-images.githubusercontent.com/74226455/148250230-304aa2eb-55d3-4c20-8ba2-8cc209539e70.png)
![image](https://user-images.githubusercontent.com/74226455/148250281-a66e4c20-eea7-43d0-8677-db6e38b29dff.png)

Thank you.

**Comment by sapitts** [comment, marked as answer]
Hi @xchengood, 

I would recommend starting with equation 4 from the cited paper. Look at the method `calculateSlipRate` in the classes [CrystalPlasticityTwinningKalidindiUpdate](https://github.com/idaholab/moose/blob/next/modules/tensor_mechanics/src/materials/crystal_plasticity/CrystalPlasticityTwinningKalidindiUpdate.C#L147) and [CrystalPlasticityKalidindiUpdate](https://github.com/idaholab/moose/blob/next/modules/tensor_mechanics/src/materials/crystal_plasticity/CrystalPlasticityKalidindiUpdate.C#L91) for examples.

A variation of Equation 1 is implemented in `CrystalPlasticityStressUpdateBase` in the method `calculateEquivalentSlipIncrement`,line [459](https://github.com/idaholab/moose/blob/next/modules/tensor_mechanics/src/materials/crystal_plasticity/CrystalPlasticityStressUpdateBase.C#L459). In this base class we solve for the plastic velocity gradient, rather than the plastic strain rate as in the Wen et al. paper; see Equation 8 on the [ComputeMultipleCrystalPlasticityStress](https://mooseframework.inl.gov/source/materials/crystal_plasticity/ComputeMultipleCrystalPlasticityStress.html) documentation page the mathematical details.

Hope this helps,
Stephanie

**Comment by ngrilli** [comment]
Dear @xchengood 

An example of the implementation we are doing for Orowan's law (your equation [4]) in the CP model in MOOSE is here:
https://github.com/ngrilli/c_pfor_am/blob/main/src/materials/FiniteStrainCrystalPlasticityDislo.C

This is still based on the monolithic CP code,
but if you want to collaborate with us to port it to the UserObject based,
feel free to contact me.

As Stephanie mentioned, equation [1] is already implemented and normally you would not change it,
unless you need to introduce additional mechanisms other than slip.

Best Regards,
Nicolò

**Reply by sapitts** [reply]
Hi @ngrilli and @xchengood 

If/when you collaborate to port the monolithic crystal plasticity code, would you refactor the code to the `StressUpdate` based version (base class [here](https://github.com/idaholab/moose/blob/next/modules/tensor_mechanics/src/materials/crystal_plasticity/CrystalPlasticityStressUpdateBase.C) ), and not the User Object version, please? The `StressUpdate` version has several robustness and speed advantages, which @dewenyushu has shown elsewhere. Additionally, it will be more straightforward to refactor the monothlithic-based code to the `StressUpdate` version since both versions are solely `Material` classes.

Thanks!
Stephanie

**Reply by ngrilli** [reply]
Dear @sapitts 
Thank you very much for the information, I was not aware of that.
I will definitely use the StressUpdate class.
Best Regards,
Nicolò

**Reply by xchengood** [reply]
@sapitts and @ngrilli Thank you very much for your information. It is much helpful for me to achieve this model in Moose. Since I have less experience in the development of Moose, I am still struggling with Moose study. Here, could you help me look at the following equation (10) that I do have no idea how to express it in Moose?
![image](https://user-images.githubusercontent.com/74226455/149969362-39f8ec36-ef73-435d-ae93-5bec44209d88.png)
The referenced paper is "Crystal plasticity-based creep model for solution-strengthened nickel-based alloys" (available at Chakraborty P, Jiang W. Crystal plasticity-based creep model for solution-strengthened nickel-based alloys. International Journal of Materials and Structural Integrity. 2019;13(1-3):144-59.)

**Reply by sapitts** [reply]
@xchengood sorry for the late reply, particularly since at this point you may have already solved this issue. 

I'd recommend reusing `calculateSchmidTensor` in the `CrystalPlasticityStressUpdateBase` class and passing in the array for the m^a_o vector for both the the `plane_normal_vector` and `direction_vector` arguments. Make sure to define a new tensor, too, to hold that dyadic product.

**Reply by xchengood** [reply]
@sapitts. Thank you for your suggestions very much. Appreciate.
